web/src/main/java/org/springframework/security/web/header/writers/HpkpHeaderWriter.java

```diff
@@ -0,0 +1,305 @@
+	 * @param reportOnly maps to {@link #setReportOnly(boolean)}
+	 */
+	public HpkpHeaderWriter(long maxAgeInSeconds, boolean includeSubDomains, boolean reportOnly) {
+		super();
+		this.maxAgeInSeconds = maxAgeInSeconds;
+		this.includeSubDomains = includeSubDomains;
+		this.reportOnly = reportOnly;
+		updateHpkpHeaderValue();
+	}
+
+	/**
+	 * Creates a new instance
+	 *
+	 * @param maxAgeInSeconds maps to {@link #setMaxAgeInSeconds(long)}
+	 * @param includeSubDomains maps to {@link #setIncludeSubDomains(boolean)}
+	 */
+	public HpkpHeaderWriter(long maxAgeInSeconds, boolean includeSubDomains) {
+		this(maxAgeInSeconds, includeSubDomains, true);
+	}
+
+	/**
+	 * Creates a new instance
+	 *
+	 * @param maxAgeInSeconds maps to {@link #setMaxAgeInSeconds(long)}
+	 */
+	public HpkpHeaderWriter(long maxAgeInSeconds) {
+		this(maxAgeInSeconds, false);
+	}
+
+	/**
+	 * Creates a new instance
+	 */
+	public HpkpHeaderWriter() {
+		this(DEFAULT_MAX_AGE_SECONDS);
+	}
+
+	/*
+	 * (non-Javadoc)
+	 * 
+	 * @see org.springframework.security.web.headers.HeaderWriter#writeHeaders(javax
+	 * .servlet.http.HttpServletRequest, javax.servlet.http.HttpServletResponse)
+	 */
+	public void writeHeaders(HttpServletRequest request, HttpServletResponse response) {
+		if (requestMatcher.matches(request)) {
+			if (!pins.isEmpty()) {
+				response.setHeader(reportOnly ? HPKP_RO_HEADER_NAME : HPKP_HEADER_NAME, hpkpHeaderValue);
+			} if (logger.isDebugEnabled()) {
+				logger.debug("Not injecting HPKP header since there aren't any pins");
+			}
+		}
+		else if (logger.isDebugEnabled()) {
+			logger.debug("Not injecting HPKP header since it wasn't a secure connection");
+		}
+	}
+
+	/**
+	 * <p>
+	 * Sets the value for the pin- directive of the Public-Key-Pins header.
+	 * </p>
+	 *
+	 * <p>
+	 * The pin directive specifies a way for web host operators to indicate
+	 * a cryptographic identity that should be bound to a given web host.
+	 * See <a href="http://tools.ietf.org/html/rfc7469#section-2.1.1">Section 2.1.1</a> for additional details.
+	 * </p>
+	 *
+	 * @param pins the map of base64-encoded SPKI fingerprint &amp; cryptographic hash algorithm pairs.
+	 * @throws IllegalArgumentException if pins is null
+	 */
+	public void setPins(Map<String, String> pins) {
+		Assert.notNull(pins, "pins cannot be null");
+		this.pins = pins;
+		updateHpkpHeaderValue();
+	}
+
+	/**
+	 * <p>
+	 * Adds a value for the pin- directive of the Public-Key-Pins header.
+	 * </p>
+	 *
+	 * <p>
+	 * The pin directive specifies a way for web host operators to indicate
+	 * a cryptographic identity that should be bound to a given web host.
+	 * See <a href="http://tools.ietf.org/html/rfc7469#section-2.1.1">Section 2.1.1</a> for additional details.
+	 * </p>
+	 *
+	 * @param pin the base64-encoded SPKI fingerprint.
+	 * @param algorithm the cryptographic hash algorithm.
+	 * @throws IllegalArgumentException if algorithm or pin is null
+	 */
+	public void addPin(String pin, String algorithm) {
```

Comment:
I would go as far as to say we should change this method to addSha256Pins(String... pins) since the only currently supported pin is sha256 and the setPins method would provide more flexibility if they need additional pin types and we haven't had a chance to update.